kmk/extensions/oled.py

```diff
@@ -0,0 +1,291 @@
+        )
+        self._split = split
+
+        make_key(
+            names=('OLED_BRI',), on_press=self._oled_bri, on_release=handler_passthrough
+        )
+        make_key(
+            names=('OLED_BRD',), on_press=self._oled_brd, on_release=handler_passthrough
+        )
+
+    def render_oled(self, layer, *args, **kwargs):
+        splash = displayio.Group()
+
+        for view in self._views:
+            if self._dark is False:
+                if view[4] == layer or view[4] is None:
+                    if view[5] == OledEntryType.TXT:
+                        splash.append(
+                            label.Label(
+                                terminalio.FONT,
+                                text=view[0],
+                                color=0xFFFFFF if not view[10] else 0x000000,
+                                background_color=0x000000 if not view[10] else 0xFFFFFF,
+                                anchor_point=(view[6], view[7]),
+                                anchored_position=(view[1], view[2]),
+                                label_direction=view[8],
+                                line_spacing=view[9],
+                                padding_left=1,
+                            )
+                        )
+                    elif view[5] == OledEntryType.IMG:
+                        splash.append(
+                            displayio.TileGrid(
+                                view[0],
+                                pixel_shader=view[0].pixel_shader,
+                                x=view[1],
+                                y=view[2],
+                            )
+                        )
+        gc.collect()
+        self._display.show(splash)
+
+    def updateOLED(self, sandbox):
+        self.render_oled(sandbox.active_layers[0])
+        gc.collect()
+
+    def on_runtime_enable(self, sandbox):
+        return
+
+    def on_runtime_disable(self, sandbox):
+        return
+
+    def during_bootup(self, keyboard):
+        if self._split is not None:
+            if self._split.split_side == 1:
+                self.split_side = 'L'
+                self._flip = self._flip_left
+            elif self._split.split_side == 2:
+                self.split_side = 'R'
+                self._flip = self._flip_right
+        else:
+            self.split_side = None
+
+        for idx, view in enumerate(self._views):
+            if view[3] != self.split_side and view[3] is not None:
+                del self._views[idx]
+
+        displayio.release_displays()
+        i2c = busio.I2C(keyboard.SCL, keyboard.SDA)
+        self._display = adafruit_displayio_ssd1306.SSD1306(
+            displayio.I2CDisplay(i2c, device_address=self._device_address),
+            width=self._width,
+            height=self._height,
+            rotation=180 if self._flip else 0,
+            brightness=self._brightness,
+        )
+        self.render_oled(0)
```

Comment:
Is `during_bootup` the right place to start rendering? Shouldn't we wait for the rest of the firmware to be initialized?

Reply:
If fits the following criteria:
Runs without our input so that we render the initial state of OLED without waiting for changes.
Runs only once so we aren't drawing initial state to immediately change it.

I could put it in ```before_matrix_scan``` with some variable that flips to true after the initial state has been rendered but then it will run additional check on each cycle. Am I missing something?

Reply:
I see, it only updates on layer changes.

One thing though: `0` isn't necessarily the active layer on boot.

Reply:
You're right. Defaulting prev. layer to ```None``` allows me to remove the first rendering from bootup.